File: src/core/tree.jl
"""
Search tree data structures for online POMDP planning.

The tree alternates between HistoryNode (belief nodes) and ActionNode (action edges):
  HistoryNode → (pick action) → ActionNode → (observe) → HistoryNode → ...

Each HistoryNode stores particles defining the belief and S_in (sampled state support).
Each ActionNode stores Q-values and Z_in (sampled observation support).

Every node gets a unique integer `id` at creation time (via `next_node_id!`).
Events emitted during simulation reference nodes by these ids.
"""

# --- Node ID allocator ---

"""Mutable counter producing unique sequential node ids. Reset per planner call."""
mutable struct NodeIdCounter
    next::Int
end

NodeIdCounter() = NodeIdCounter(0)

"""Return a fresh id and advance the counter."""
function next_node_id!(counter::NodeIdCounter)
    counter.next += 1
    return counter.next
end

# --- ActionNode ---

"""
    ActionNode

Represents an action edge in the search tree.

Fields:
- `id`: unique integer id, assigned at creation
- `children`: observation → child HistoryNode
- `Z_in`: set of observations seen (sampled observation support)
- `N`: visit count
- `Q_nominal`: average return from simulations (standard POMCP value)
- `Q_robust`: robust value from worst-case backup (computed by RobustBackup)
- `dirty`: whether this node needs robust value recomputation
"""
mutable struct ActionNode
    id::Int
    children::Dict{Int, Any}    # obs → HistoryNode (Any to handle forward reference)
    Z_in::Set{Int}
    N::Int
    Q_nominal::Float64
    Q_robust::Float64
    dirty::Bool
end

function ActionNode(id::Int)
    return ActionNode(id, Dict{Int, Any}(), Set{Int}(), 0, 0.0, 0.0, true)
end

# --- HistoryNode ---

"""
    HistoryNode

Represents a belief node in the search tree.

Fields:
- `particles`: list of states sampled at this node (may contain duplicates)
- `S_in`: set of unique states (sampled state support)
- `children`: action → ActionNode
- `N`: visit count
- `depth`: depth in tree (root = 0)
- `dirty`: whether this node needs robust value recomputation
"""
mutable struct HistoryNode
    id::Int
    particles::Vector{Int}
    S_in::Set{Int}
    children::Dict{Int, ActionNode}
    N::Int
    depth::Int
    dirty::Bool
    V_robust::Float64       # cached max_a Q_robust(a), updated during robust backup
    V_nominal::Float64      # cached max_a Q_nominal(a), updated during simulation
    V_rollout::Float64      # rollout return saved when this node is first expanded
    has_rollout::Bool       # whether V_rollout has been set
end

function HistoryNode(id::Int, depth::Int)
    return HistoryNode(id, Int[], Set{Int}(), Dict{Int, ActionNode}(), 0, depth, true, 0.0, 0.0, 0.0, false)
end

# Fix the forward reference: ActionNode children actually hold HistoryNodes
# (Julia handles this fine since we used Any in ActionNode.children)

# --- Mutating helpers ---

"""Add a particle (sampled state) to a history node. Marks node as dirty."""
function add_particle!(node::HistoryNode, s::Int)
    push!(node.particles, s)
    push!(node.S_in, s)
    node.N += 1
    node.dirty = true
end

"""
Record a new observation at an action node. Creates child if needed. Marks as dirty.
Returns (child_node, was_created) so callers can log which nodes were new.
"""
function get_or_create_child!(action_node::ActionNode, z::Int, child_depth::Int,
                              id_counter::NodeIdCounter)
    push!(action_node.Z_in, z)
    was_created = false
    if !haskey(action_node.children, z)
        new_id = next_node_id!(id_counter)
        action_node.children[z] = HistoryNode(new_id, child_depth)
        action_node.dirty = true
        was_created = true
    end
    return (action_node.children[z]::HistoryNode, was_created)
end

"""
Expand a history node: create ActionNodes for all actions.
Returns (created_ids, created_actions) where both vectors align:
  created_ids[i] is the new ActionNode's id
  created_actions[i] is the action index (1..n_actions) that node corresponds to
"""
function expand!(node::HistoryNode, n_actions::Int, id_counter::NodeIdCounter)
    created_ids = Int[]
    created_actions = Int[]
    for a in 1:n_actions
        if !haskey(node.children, a)
            new_id = next_node_id!(id_counter)
            node.children[a] = ActionNode(new_id)
            push!(created_ids, new_id)
            push!(created_actions, a)
        end
    end
    return (created_ids, created_actions)
end

"""Check if a history node is a leaf (not yet expanded)."""
is_leaf(node::HistoryNode) = isempty(node.children)

"""
Update the cached V_robust from children's Q_robust values.

Only considers actions that have been visited (N > 0). If no action has been
visited, falls back to V_rollout (the sampled rollout return at this node).
If V_rollout hasn't been set either, falls back to 0.
"""
function update_robust_value!(node::HistoryNode)
    if is_leaf(node)
        node.V_robust = node.has_rollout ? node.V_rollout : 0.0
        return
    end

    visited_qs = [an.Q_robust for (_, an) in node.children if an.N > 0]

    if isempty(visited_qs)
        node.V_robust = node.has_rollout ? node.V_rollout : 0.0
    else
        node.V_robust = maximum(visited_qs)
    end
end

"""
Update the cached V_nominal from children's Q_nominal values.

Same logic as update_robust_value!: only visited actions count; V_rollout is the fallback.
"""
function update_nominal_value!(node::HistoryNode)
    if is_leaf(node)
        node.V_nominal = node.has_rollout ? node.V_rollout : 0.0
        return
    end

    visited_qs = [an.Q_nominal for (_, an) in node.children if an.N > 0]

    if isempty(visited_qs)
        node.V_nominal = node.has_rollout ? node.V_rollout : 0.0
    else
        node.V_nominal = maximum(visited_qs)
    end
end

File: src/robust_pomdp/core/tree.py
"""
Search tree data structures for online POMDP planning.

The tree alternates between HistoryNode (belief nodes) and ActionNode (action edges):
  HistoryNode -> (pick action) -> ActionNode -> (observe) -> HistoryNode -> ...

Each HistoryNode stores particles defining the belief and S_in (sampled state support).
Each ActionNode stores Q-values and Z_in (sampled observation support).

Every node gets a unique integer id at creation time (via NodeIdCounter.next_id()).
Events emitted during simulation reference nodes by these ids.

Action / state / observation ids are 0-based throughout (Python convention).
"""

from __future__ import annotations


class NodeIdCounter:
    """Counter producing unique sequential node ids. Reset per planner call."""

    def __init__(self) -> None:
        self.next: int = 0

    def next_id(self) -> int:
        """Return a fresh id and advance the counter."""
        self.next += 1
        return self.next


class ActionNode:
    """An action edge in the search tree.

    Attributes:
        id: unique integer id, assigned at creation.
        children: observation -> child HistoryNode.
        Z_in: set of observations seen (sampled observation support).
        N: visit count.
        Q_nominal: average return from simulations (standard POMCP value).
        Q_robust: robust value from worst-case backup.
        dirty: whether this node needs robust value recomputation.
    """

    def __init__(self, id: int) -> None:
        self.id = id
        self.children: dict[int, HistoryNode] = {}
        self.Z_in: set[int] = set()
        self.N: int = 0
        self.Q_nominal: float = 0.0
        self.Q_robust: float = 0.0
        self.dirty: bool = True

    def get_or_create_child(self,
                            z: int,
                            child_depth: int,
                            id_counter: NodeIdCounter
                            ) -> tuple[HistoryNode, bool]:
        """Record observation z and return its child HistoryNode (creating if needed).

        Marks the action node as dirty when a new child is created.
        Returns (child_node, was_created).
        """
        self.Z_in.add(z)
        was_created = False
        if z not in self.children:
            new_id = id_counter.next_id()
            self.children[z] = HistoryNode(new_id, child_depth)
            self.dirty = True
            was_created = True
        return self.children[z], was_created


class HistoryNode:
    """A belief node in the search tree.

    Attributes:
        id: unique integer id.
        depth: depth in tree (root = 0).
        particles: list of states sampled at this node (may contain duplicates).
        S_in: set of unique states (sampled state support).
        children: action -> ActionNode.
        N: visit count.
        dirty: whether this node needs robust value recomputation.
        V_robust: cached max_a Q_robust(a), updated during robust backup.
        V_nominal: cached max_a Q_nominal(a), updated during simulation.
        V_rollout: rollout return saved when this node was first expanded.
        has_rollout: whether V_rollout has been set.
    """

    def __init__(self, id: int, depth: int) -> None:
        self.id = id
        self.depth = depth
        self.particles: list[int] = []
        self.S_in: set[int] = set()
        self.children: dict[int, ActionNode] = {}
        self.N: int = 0
        self.dirty: bool = True
        self.V_robust: float = 0.0
        self.V_nominal: float = 0.0
        self.V_rollout: float = 0.0
        self.has_rollout: bool = False

    def add_particle(self, s: int) -> None:
        """Add a sampled state to this node. Marks node as dirty."""
        self.particles.append(s)
        self.S_in.add(s)
        self.N += 1
        self.dirty = True

    def is_leaf(self) -> bool:
        """Return True if this node has not yet been expanded."""
        return not self.children

    def expand(self, n_actions: int, id_counter: NodeIdCounter) -> tuple[list[int], list[int]]:
        """Create ActionNodes for any actions not yet expanded.

        Returns (created_ids, created_actions): aligned lists where created_ids[i]
        is the new ActionNode's id and created_actions[i] is the action index (0-based).
        """
        created_ids: list[int] = []
        created_actions: list[int] = []
        for a in range(n_actions):
            if a not in self.children:
                new_id = id_counter.next_id()
                self.children[a] = ActionNode(new_id)
                created_ids.append(new_id)
                created_actions.append(a)
        return created_ids, created_actions

    def update_robust_value(self) -> None:
        """Update cached V_robust from children's Q_robust values.

        Only considers actions that have been visited (N > 0). If no action has been
        visited, falls back to V_rollout. If V_rollout hasn't been set either,
        falls back to 0.0. (Bug fix from Julia Session 5: unvisited Q_robust=0 was
        contaminating parent backups.)
        """
        if self.is_leaf():
            self.V_robust = self.V_rollout if self.has_rollout else 0.0
            return

        visited_qs = [an.Q_robust for an in self.children.values() if an.N > 0]

        if not visited_qs:
            self.V_robust = self.V_rollout if self.has_rollout else 0.0
        else:
            self.V_robust = max(visited_qs)

    def update_nominal_value(self) -> None:
        """Update cached V_nominal from children's Q_nominal values.

        Same fallback logic as update_robust_value.
        """
        if self.is_leaf():
            self.V_nominal = self.V_rollout if self.has_rollout else 0.0
            return

        visited_qs = [an.Q_nominal for an in self.children.values() if an.N > 0]

        if not visited_qs:
            self.V_nominal = self.V_rollout if self.has_rollout else 0.0
        else:
            self.V_nominal = max(visited_qs)
